cpp/src/arrow/record_batch.h

```diff
@@ -300,6 +311,11 @@ class ARROW_EXPORT RecordBatchReader {
   /// \brief finalize reader
   virtual Status Close() { return Status::OK(); }
 
+  /// \brief Get the device type for record batches this reader produces
+  ///
+  /// default implementation is to return ARROW_DEVICE_CPU
+  virtual DeviceAllocationType device_type() const { return DeviceAllocationType::kCPU; }
```

Comment:
This is going to be wrong if we import a RecordBatchReader from a non-CPU ArrowDeviceArrayStream? Or is that right now not yet possible?

Reply:
As of yet, I haven't implemented figuring out what type to send here. Since we aren't storing the device type at the record batch level, this would need to spin through the buffers of its columns which could *potentially* be from multiple devices if a user did something weird. So I'm not sure what the best solution here is. I'd rather not add a `RecordBatch` level device_type member, but I don't yet know how to best handle the situation if a consumer does something bad

Reply:
Well, this is an abstract class, so there's no constructor here :-)

Reply:
So the suggestion is we *should* add a `device_type` member to the `RecordBatch` class and have it be provided?

Reply:
I've propagated the device_type throughout the readers now and to the static `Make` methods for constructing the readers, so now this should be correct.